sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangeStreamErrorTest.java

```diff
@@ -0,0 +1,497 @@
+    serviceHelper.reset();
+    serviceHelper.stop();
+    mockSpannerService.reset();
+    resetDaoFactoryFields();
+  }
+
+  @Test
+  public void testResourceExhaustedDoesNotRetry() {
+    mockSpannerService.setExecuteStreamingSqlExecutionTime(
+        SimulatedExecutionTime.ofStickyException(Status.RESOURCE_EXHAUSTED.asRuntimeException()));
+
+    final Timestamp now = Timestamp.now();
+    final Timestamp after3Seconds =
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds() + 3, now.getNanos());
+    try {
+      pipeline.apply(
+          SpannerIO.readChangeStream()
+              .withSpannerConfig(getSpannerConfig())
+              .withChangeStreamName(TEST_CHANGE_STREAM)
+              .withMetadataDatabase(TEST_DATABASE)
+              .withInclusiveStartAt(now)
+              .withInclusiveEndAt(after3Seconds));
+      pipeline.run().waitUntilFinish();
+    } finally {
+      thrown.expect(PipelineExecutionException.class);
+      thrown.expectMessage(ErrorCode.RESOURCE_EXHAUSTED.name());
+    }
+  }
+
+  @Test
+  public void testUnavailableExceptionRetries() {
+    mockSpannerService.setExecuteStreamingSqlExecutionTime(
+        SimulatedExecutionTime.ofExceptions(
+            ImmutableSet.of(
+                Status.UNAVAILABLE.asRuntimeException(),
+                Status.RESOURCE_EXHAUSTED.asRuntimeException())));
+
+    final Timestamp now = Timestamp.now();
+    final Timestamp after3Seconds =
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds() + 3, now.getNanos());
+    try {
+      pipeline.apply(
+          SpannerIO.readChangeStream()
+              .withSpannerConfig(getSpannerConfig())
+              .withChangeStreamName(TEST_CHANGE_STREAM)
+              .withMetadataDatabase(TEST_DATABASE)
+              .withInclusiveStartAt(now)
+              .withInclusiveEndAt(after3Seconds));
+      pipeline.run().waitUntilFinish();
+    } finally {
+      assertThat(
+          mockSpannerService.countRequestsOfType(ExecuteSqlRequest.class), Matchers.greaterThan(1));
+      thrown.expect(PipelineExecutionException.class);
+      thrown.expectMessage(ErrorCode.RESOURCE_EXHAUSTED.name());
+    }
+  }
+
+  @Test
+  public void testAbortedExceptionRetries() {
+    mockSpannerService.setExecuteStreamingSqlExecutionTime(
+        SimulatedExecutionTime.ofStickyException(Status.ABORTED.asRuntimeException()));
+
+    final Timestamp now = Timestamp.now();
+    final Timestamp after3Seconds =
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds() + 3, now.getNanos());
+    try {
+      pipeline.apply(
+          SpannerIO.readChangeStream()
+              .withSpannerConfig(getSpannerConfig())
+              .withChangeStreamName(TEST_CHANGE_STREAM)
+              .withMetadataDatabase(TEST_DATABASE)
+              .withInclusiveStartAt(now)
+              .withInclusiveEndAt(after3Seconds));
+      pipeline.run().waitUntilFinish();
+    } finally {
+      assertThat(
```

Comment:
Today this value is always zero. It seems very sensitive to internal details. Can we change it so that the assertion is checking correct results?